lib/vm/src/instance/mod.rs

```diff
@@ -861,11 +821,19 @@ impl Instance {
 
         if let Ok(mut ret) = ret {
             if ret == 0 {
-                ret = self.do_wait(memory_index.as_u32(), dst, timeout);
-            }
-            if ret == 0xffff {
-                // ret is 0xffff if there is more than 2^32 waiter in queue
-                return Err(Trap::lib(TrapCode::TableAccessOutOfBounds));
+                let memory = self.get_local_vmmemory_mut(memory_index);
+                let location = NotifyLocation { address: dst };
+                let timeout = if timeout < 0 {
+                    None
+                } else {
+                    Some(std::time::Duration::from_nanos(timeout as u64))
+                };
+                let waiter = memory.do_wait(location, timeout);
+                if waiter.is_none() {
+                    // ret is None if there is more than 2^32 waiter in queue or some other error
+                    return Err(Trap::lib(TrapCode::TableAccessOutOfBounds));
```

Comment:
We need to make sure the behaviour here matches the spec, and that we trap with an appropriate code (for allowing proper debugging).

See some related comments below.

Reply:
Note: this currently matches the spec, because it only returns `None` if the limit is exceeded, but this would be pretty easy to mess up in a future refactor, and we also want to add trait extensibility.

Reply:
I change to return a Result with custom error instead @theduke . Is that better now?

Reply:
Yes!